### PR TITLE
provide the ability to include additional scripts in lazybones scripts

### PR DESCRIPTION
**Pull request by drmaas**
Example (from test):

```
def subProperties = include("sub1.groovy", ["maxThreads": maxThreads])
```

### CODE REVIEW

**Comment by pledbrook** [comment]
Thanks for the contribution! It's a fairly substantial addition in terms of functionality, so it may take a little time for me to review (I gave it a quick once over, but realised it needed more attention). In the meantime, would you mind fixing the CodeNarc violations?

**Comment by drmaas** [comment]
Sure thing.

**Comment by drmaas** [comment]
@pledbrook thoughts?

**Comment by pledbrook** [comment]
I'm sorry, I haven't been on top of things in the last month or so. I'm honestly just tired. I have a few things to sort out first, but I'll see if I can take a stab today and try to just get this done.

File: lazybones-app/src/main/groovy/uk/co/cacoethes/lazybones/LazybonesScript.groovy [hunk 1]
```diff
@@ -162,6 +168,21 @@ class LazybonesScript extends Script {
     }
 
     /**
+     * Include another script in the main script, and execute it in-line.
+     * @param file subscript
+     * @param bindings values to pass from parent script to this script
+     * @return the last expression evaluated in the subscript
+     */
+    def include(String file, Map<String, String>...bindings) {
```
Comment (pledbrook) [inline lazybones-app/src/main/groovy/uk/co/cacoethes/lazybones/LazybonesScript.groovy]: Is there a reason `binding` is a varargs? It makes more sense to me for this to be a single `Map` argument.

File: lazybones-app/src/integTest/groovy/uk/co/cacoethes/lazybones/CreateFunctionalSpec.groovy [hunk 2]
```diff
@@ -67,6 +67,25 @@ class CreateFunctionalSpec extends AbstractFunctionalSpec {
...
+                "my-app"]
+        def exitCode = runCommand(args, baseWorkDir, ["org.example", "1.0-SNAPSHOT", "4", "org.foo", "2.0-SNAPSHOT", "8"])
+
+        then: "It successfully completes"
+        exitCode == 0
+
+        and: "The generated build file contains the expected group ID and version"
+        def text = new File(baseWorkDir, "my-app/build.gradle").text.trim()
+        text.contains("group = \"org.example\"")
+        text.contains("version = \"1.0-SNAPSHOT\"")
```
Comment (pledbrook) [inline lazybones-app/src/integTest/groovy/uk/co/cacoethes/lazybones/CreateFunctionalSpec.groovy]: You should probably also check that the included script is deleted.

File: lazybones-app/src/main/groovy/uk/co/cacoethes/lazybones/LazybonesScript.groovy [hunk 3]
```diff
@@ -162,6 +168,21 @@ class LazybonesScript extends Script {
     }
 
     /**
+     * Include another script in the main script, and execute it in-line.
+     * @param file subscript
+     * @param bindings values to pass from parent script to this script
```
Comment (pledbrook) [inline lazybones-app/src/main/groovy/uk/co/cacoethes/lazybones/LazybonesScript.groovy]: This should make it clear that we're talking about _additional_ model parameters on top of those available to the parent script.

File: lazybones-app/src/main/groovy/uk/co/cacoethes/lazybones/commands/InstallationScriptExecuter.groovy [hunk 4]
```diff
@@ -55,15 +66,19 @@ class InstallationScriptExecuter {
...
+    Script runPostInstallScript(String scriptFileName,
+                                List tmplQualifiers,
+                                File targetDir,
+                                File templateDir,
+                                Map<String, String> model) {
+        def installScriptFile = new File(templateDir, scriptFileName)
         if (installScriptFile.exists()) {
             def script = initializeScript(model, tmplQualifiers, installScriptFile, targetDir, templateDir)
-            script.run()
+            scriptReturnValue = script.run()
```
Comment (pledbrook) [inline lazybones-app/src/main/groovy/uk/co/cacoethes/lazybones/commands/InstallationScriptExecuter.groovy]: This should not be internal state of the executor. Why not return the script value directly from `runPostInstallScript()`? I notice that the only thing that uses the `Script` object is a test. Thinking about it, the return value of the script makes more sense as a return value of a `run` method. We could modify the test to use `initializeScript()` instead perhaps.

If we absolutely need the `Script` object returned as well, then we can introduce a new type that has properties for the `Script` and the script's return value after execution. Heavyweight perhaps, but correct.

**Comment by drmaas** [comment]
thanks @pledbrook - hopefully I'll get some time in the next week to address your comments.

File: lazybones-app/src/main/groovy/uk/co/cacoethes/lazybones/LazybonesScript.groovy (hunk 1 above)
Comment (drmaas) [inline lazybones-app/src/main/groovy/uk/co/cacoethes/lazybones/LazybonesScript.groovy]: Not particularly, will address.

File: lazybones-app/src/main/groovy/uk/co/cacoethes/lazybones/commands/InstallationScriptExecuter.groovy (hunk 4 above)
Comment (drmaas) [inline lazybones-app/src/main/groovy/uk/co/cacoethes/lazybones/commands/InstallationScriptExecuter.groovy]: Opting for the former, as it is simpler.

**Comment by drmaas** [comment]
@pledbrook addressed comments, PTAL

**Comment by pledbrook** [comment]
Thanks for the contribution! It won't go into a new release until documented though. Hopefully won't be long.

**Pull request merged**